lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,111 @@
+-- |
+-- Module      : Amazonka.Auth.SSO
+-- Copyright   : (c) 2013-2021 Brendan Hay
+-- License     : Mozilla Public License, v. 2.0.
+-- Maintainer  : Brendan Hay <brendan.g.hay+amazonka@gmail.com>
+-- Stability   : provisional
+-- Portability : non-portable (GHC extensions)
+module Amazonka.Auth.SSO where
+
+import Amazonka.Auth.Background (fetchAuthInBackground)
+import Amazonka.Auth.Exception
+import qualified Amazonka.Crypto as Crypto
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (ISO8601, Time (..))
+import Amazonka.Env (Env, Env' (..))
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens ((^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Exception.Lens (handling_, _IOException)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+fromSSO ::
+  (MonadIO m, Foldable withAuth) =>
+  FilePath ->
+  Region ->
+  Text ->
+  Text ->
+  Env' withAuth ->
+  m Env
+fromSSO cache ssoRegion accountId roleName env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  pure $ env {envAuth = Identity auth}
```

Comment:
Question: It's correct to not update region on the returned environment, right? This is because the SSO region and the region the library user wants might be different. Even though we respect `AWS_REGION` somewhat, we might also have the region we want in the input `Env' withAuth`.

If I've got things right, would you mind adding a comment here? This seems a bit subtle.

Reply:
I added a comment on the use of `ssoRegion` that I think addresses this. Going to optimistically Resolve conversation.